File: JUNHEE/프로그래머스_week9/여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<string> answer;
bool flag = false;
void DFS(string now, vector<vector<string>>& tickets, vector<bool>& visit, int size){
    answer.push_back(now);
    if(size == tickets.size()){
        flag = true;
        return;
    }
    for(int i = 0; i < tickets.size(); i++){
        if(visit[i] == false && tickets[i][0] == now){
            visit[i] = true;
            DFS(tickets[i][1], tickets, visit, size + 1);
            if(flag) return;
            visit[i] = false;
            answer.pop_back();
        }
    }
}

vector<string> solution(vector<vector<string>> tickets) {
    
    vector<bool> visit(tickets.size(), false);
    sort(tickets.begin(), tickets.end());
    DFS("ICN",tickets,visit,0);
    return answer;
}

File: DONGUK/프로그래머스_WEEK 9/여행경로.py
from collections import deque
import copy

cities = {}
ticket_cnt = {}
answer = []


def search(arr, now, visited, N):
    if len(arr) == N+1:
        arr_cp = copy.deepcopy(arr)  
        answer.append(arr_cp)
        return
    
    if now not in cities:
        return
    
    candidate = cities[now]
    candidate.sort()
    
    for next_city in candidate:
        if (now, next_city) not in visited or ticket_cnt[(now, next_city)] > 0:
            arr.append(next_city)
            visited.append((now, next_city))
            ticket_cnt[(now, next_city)] -= 1
            search(arr, next_city, visited, N)
            visited.remove((now, next_city))
            ticket_cnt[(now, next_city)] += 1
            arr.pop()

        

def solution(tickets):
    global answer
    
    for start, end in tickets:
        if start not in cities:
            cities[start] = [end]
            
        else:
            cities[start].append(end)

        if (start, end) not in ticket_cnt:
            ticket_cnt[(start, end)] = 1

        else:
            ticket_cnt[(start, end)] += 1
    
    
    search(["ICN"], "ICN", [], len(tickets))
    
    answer.sort()
    
    answer = answer[0]
    
    return answer
    

